test/python/circuit/test_unitary.py

```diff
@@ -183,3 +183,26 @@ def test_labeled_unitary(self):
         instr = qobj.experiments[0].instructions[0]
         self.assertEqual(instr.name, 'unitary')
         self.assertEqual(instr.label, 'xy')
+
+    def test_qasm_unitary(self):
+        """test that a custom unitary can be converted to qasm"""
+        qr = QuantumRegister(2, 'q0')
+        cr = ClassicalRegister(1, 'c0')
+        qc = QuantumCircuit(qr, cr)
+        matrix = numpy.array([[1, 0], [0, 1]])
+        unitary_gate = UnitaryGate(matrix, label="custom_gate")
+
+        qc.x(qr[0])
+        qc.append(unitary_gate, [qr[0]])
+        qc.append(unitary_gate, [qr[1]])
+
+        expected_qasm = "OPENQASM 2.0;\n" \
+                        "include \"qelib1.inc\";\n" \
+                        "qreg q0[2];\ncreg c0[1];\n" \
+                        "x q0[0];\n" \
+                        "gate custom_gate p0 {\n" \
+                        "\tu3(0.0,0.0,0.0) p0;\n" \
+                        "}\n" \
+                        "custom_gate q0[0];\n" \
+                        "custom_gate q0[1];\n"
+        self.assertEqual(expected_qasm, qc.qasm())
```

Comment:
Can you add a test for writing the qasm of a circuit twice? And one for decomposing a two qubit unitary?

Reply:
I have added a test for writing the qasm of a circuit twice. However the decomposition for a 2 qubit unitary seems to be somewhat random (I believe due to [this algorithm](https://github.com/Qiskit/qiskit-terra/blob/998c232b00c56040a14af23999a07787c0090e5e/qiskit/quantum_info/synthesis/two_qubit_decompose.py#L164)) and so the qasm generated each time is ever so slightly different but equivalent. This means we cannot do a straightforward equivalence check of the new qasm. We also cannot test that the original circuit and the circuit from the loaded qasm generate the same results as you cannot load custom gates into the DAG (`qiskit.exceptions.QiskitError: 'Custom non-opaque gates are not supported by as_to_dag module'`), so I'm not sure there is a good way to test this part.

Reply:
You could try a two qubit identity (`np.eye(4)`), which should reliably decompose to two `u3(0.0,0.0,0.0)`s. The issue on loading custom gates from qasm is tracked in #1566 .

Reply:
Thanks for the suggestion! Added in e0da9c7